File: Kevin van den Bekerom/DSLQL/src/main/java/org/uva/sea/ql/ast/expr/Eq.java
package org.uva.sea.ql.ast.expr;

import org.uva.sea.ql.ast.form.Context;
import org.uva.sea.ql.ast.visit.Visitable;
import org.uva.sea.ql.ast.visit.Visitor;

public class Eq extends BinaryExpr implements Visitable {	
	
	public Eq(Expr lhs, Expr rhs) {
		super.lhs = lhs;
		super.rhs = rhs;
	}
	
	//TODO: different check for Strings and booleans!!! Need to know the type
	@Override
	public Boolean eval() {
		return lhs.eval().equals(rhs.eval());
	}
	
	public void accept(Visitor visitor, Object context) {
		visitor.visit(this, context);
	}

	@Override
	public String toString() {
		return super.toString() + " (==)";
	}
	
	@Override
	public Type getType(Context context) {
		return Type.BOOLEAN;
	}
}

File: Edward Poot/src/ql/ast/expr/binaryoperatorexpr/eq.go
package binaryoperatorexpr

import "ql/ast/expr"

type Eq struct {
	BinaryOperator
}

func NewEq(lhs expr.Expr, rhs expr.Expr) Eq {
	return Eq{BinaryOperator{Lhs: lhs, Rhs: rhs}}
}

func (e Eq) Eval(s interface{}) interface{} {
	switch e.Lhs.Eval(s).(type) {
	case int:
		return e.Lhs.Eval(s).(int) == e.Rhs.Eval(s).(int)
	case bool:
		return e.Lhs.Eval(s).(bool) == e.Rhs.Eval(s).(bool)
	}

	return nil
}
